src/Featurizers/LagLeadOperatorFeaturizer.h

```diff
@@ -0,0 +1,234 @@
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
+
+#include "../Featurizer.h"
+#include "../Archive.h"
+#include "Components/PipelineExecutionEstimatorImpl.h"
+#include "Components/FilterDecoratorFeaturizer.h"
+#include "Components/GrainFeaturizerImpl.h"
+#include "Components/WindowFeaturizerBase.h"
+
+namespace Microsoft {
+namespace Featurizer {
+namespace Featurizers {
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         LagLeadOperatorTransformer
+///  \brief         Copy values from prior data or future data
+///
+template <typename InputTupleT>
+class LagLeadOperatorTransformer :
+    public Transformer<
+        InputTupleT,
+        Microsoft::Featurizer::ColMajMatrix<std::double_t>
+    > {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    static_assert(Components::Details::IsTuple<InputTupleT>::value, "'InputTupleT' must be a tuple");
+
+    using BaseType =
+        Transformer<
+            InputTupleT,
+            Microsoft::Featurizer::ColMajMatrix<std::double_t>
+        >;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    LagLeadOperatorTransformer(std::uint32_t horizon, std::vector<std::int64_t> offsets);
+    LagLeadOperatorTransformer(Archive &ar);
+
+    ~LagLeadOperatorTransformer(void) override = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(LagLeadOperatorTransformer);
+
+    bool operator==(LagLeadOperatorTransformer const &other) const;
+    bool operator!=(LagLeadOperatorTransformer const &other) const;
+
+    void save(Archive &ar) const override;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Types
+    // |
+    // ----------------------------------------------------------------------
+    using GrainType  = std::vector<std::string>;
+    // TODO: change target type to general types
+    using TargetType = std::double_t;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    Components::CircularBuffer<TargetType>  _buffer;
+    std::int64_t const                      _max_future_offset;
+    std::int64_t const                      _max_prior_offset;
+    std::uint32_t const                     _horizon;
+    std::vector<std::int64_t> const         _offsets;
```

Comment:
>std::vector<std::int64_t> const   [](start = 4, length = 33)

This can not be moved if it is const and so the fact that is copy constructor is prohibited becomes irrelevant, everything is being copied.

Reply:
This vector is only copied in the constructor so it may not cost that much. Do you have any suggestions on this?

Reply:
I suggest phibiting both copy and move for the transformer

---
In reply to: [394597621](https://github.com/microsoft/FeaturizersLibrary/pull/171#discussion_r394597621) [](ancestors = 394597621)